basic-api.yml

```diff
@@ -0,0 +1,142 @@
+openapi: 3.0.0
+info:
+  description: This is an open HTTP API specification for issuing Verifiable Credentials based on the W3C Verifiable Credentials Data Model (https://www.w3.org/TR/vc-data-model/). Implementations of this API may differ in terms of which representation formats or proof formats they do or do not support. Implementations MAY choose to not support certain parts of this API (e.g. optional parts include changing and retrieving the status of a credential, or refreshing a credential).
+  version: "0.0.2"
+  title: Verifiable Credential Issuer HTTP API
+  contact:
+    email: msporny@digitalbazaar.com
+  license:
+    name: All documents in this Repository are licensed by contributors under the [W3C Software and Document License](http://www.w3.org/Consortium/Legal/copyright-software).
+paths:
+  /credentials:
```

Comment:
Missing "internal" tag

Reply:
This API proposes that there is no such thing as an internal vs. external HTTP API. There is just an HTTP API.

The way that is determined is via authz to the endpoint. If you don't want the API to be available to the public Internet, put it behind a firewall. If it's on the public internet, make sure it has appropriate authz to access the endpoint's functionality.

Does that work for you, @llorllale?

Reply:
I have a problem with that. I don't believe that externally facing APIs for issuance and verification should have the same interaction patterns as internally facing ones. This is not an authz issue; it's a matter of whether the industry standardizes something that's capable of working peer-to-peer. I don't mind someone using an API externally; I do mind claiming that that's a standard. What we are standardizing is the use of the API for internal-facing use cases, and explicitly NOT the use of the API in external-facing cases.

Reply:
I agree with @dhh1128 if only because it simplifies the analysis by a ton.

Reply:
See https://github.com/w3c-ccg/vc-issuer-http-api/issues/1 for an issue on removing the (current) public APIs.

Reply:
> I don't believe that externally facing APIs for issuance and verification should have the same interaction patterns as internally facing ones. This is not an authz issue; it's a matter of whether the industry standardizes something that's capable of working peer-to-peer.

+1 to an API that works with existing infrastructure and patterns as well as one that supports newer peer-to-peer models. However, instead of marking this API as "external" or "internal" (which I personally don't think really does anything *in practice* -- and I could elaborate on as needed), I would prefer to see a proposal to add something to this API that would ensure we can achieve both goals. Perhaps the request data could include additional information that, when present, would be used to deliver the issued credential directly to a "peer", as opposed to returning it in the response whereby another component of the issuer application would then hand it off in some other fashion (separation of concerns).

Reply:
+1 to removing the confusing 'internal' / 'external' (or public/private, whatever) distinction and coming up with clearer terms for API intension.  We could have "issuer" endpoints and "holder" endpoints?

Question: in a peer-to-peer scenario, is the HTTP API still relevant? (just making sure as currently peer-to-peer use cases I have seen make no assumption of web services...)

Reply:
I agree with @msporny, I think we should not care how you make the API internal, thats out of scope here... you could use an API gateway like kong, cloudflare / firewall rules, OAuth Tokens, Macaroons... or you can host it unprotected on some network you think is not publicly accessible... I feel like that last one is a bad idea for something that makes credentials... we should be encouraging the API to be publicly accessible, but secured, such that only authorized parties can access it (internal).

Reply:
>+1 to an API that works with existing infrastructure and patterns as well as one that supports newer peer-to-peer models. However, instead of marking this API as "external" or "internal" (which I personally don't think really does anything in practice -- and I could elaborate on as needed), I would prefer to see a proposal to add something to this API that would ensure we can achieve both goals.

The internal/external distinction was a compromise. If we now want to walk that back, then I'll bring up my original concern again.

I am emphatically opposed to using this API for external-facing issuance *as a standard*. Such a standard would enthrone for the entire industry the assumption that issuers are servers and holders are clients. This preserves a power imbalance between issuers and holders, institutions and ordinary people, that is counter to the goals of SSI. I want Alice, an ordinary human being, to be able to issue a credential to her babysitter. Alice cannot be expected to run a web server to do that. Nor can she be expected to contract with some company to do it for her.

I didn't care about this distinction when we were saying this was an API for internal clients to call within the issuer's infrastructure, because if issuers want to use a client-server paradigm internally, that's great. But externally, I care a lot.

Client-server and Web API models can be done inside peer-to-peer, but not the other way around. There's no "make the API work for both" when the very fact that it's a web API is what doesn't work.

Reply:
@dhh1128 wrote: 
> I want Alice, an ordinary human being, to be able to issue a credential to her babysitter. Alice cannot be expected to run a web server to do that. Nor can she be expected to contract with some company to do it for her.

+1 -- that's a hard requirement, IMHO.

I don't understand why we're coming to different conclusions, though. 

I believe what you're saying is: "I'm concerned that the existence of a Issuer HTTP API as a standard, that could be used in a public-facing server-to-client setting, will send a message that this ecosystem doesn't allow for an individual to issue a credential to another individual."

Is that what you're saying and/or concerned about, @dhh1128?

Reply:
Yes.

Standardizing the institution-centric server API first, as a separate goal, because that's easy and obvious, will force us to have a *different* and *later* answer for the use cases that are individual-centric. And institutions won't care about that second answer, when we finally get around to thinking about them, because it's not one of their use cases. They'll spend zero dollars/euros/rubles on that standard.

If we want to preserve SSI as a thing WRT VCs, we must make the first issuance standard one that's usable for Alice, because we will never get another chance to help her. It should *also* be usable for institutions (the ones who are investing right now), for sure. But it can't only be usable for them.

When I heard that we were considering a RESTful API for issuers, I immediately raised my concerns on [a long thread in the CCG](https://lists.w3.org/Archives/Public/public-credentials/2019Dec/0072.html). I thought that we concluded that my concerns were ignorable because the API we were building was intended to be called by internal clients at an issuer -- not by prospective external clients. If I misunderstood, then let's go revisit that conversation on email or in a meeting, because the scope of this particular PR is probably not the right one.

Reply:
To my mind the first interface that should be defined is between the application services and the subject database. The rationale for this is that all issuing organisations (IdPs) already have subject databases, but dont have any VC issuing systems. So the first interface that needs to be built is between these two.

Reply:
Can you make a proposal for an HTTP API that works for both the SSI and Institutional use case?

Keeping the API potentially public is actually closer to supporting this goal. 

If the objection is to the concept of an HTTP API, I'm not sure we can do much about that in this work item, but if it's about the format of a POST request, response, authorization, endpoints, thats certainly in the realm of addressable.

Concrete example: As a did subject / controller I request a credential proving my did is resolvable, the api validates the VC, checks to make sure the did is resolvable, and issues the credential and returns it to me in an HTTP response... the API uses http signature authorization protection, so that only the did controller can request a proof of resolution credential... 

Are you asserting that HTTP is incompatible with SSI?

Which part of this API is precluding the SSI use case?

Reply:
>Can you make a proposal for an HTTP API that works for both the SSI and Institutional use case?

That depends. Am I proposing an HTTP API that supports internal callers, where outside interactions are different? Or am I proposing an HTTP API that must support both internal and external callers?

What we agreed to was an API that supports internal callers. What we have here is fine for that. I don't have any objection.

If we want an HTTP API that supports external interactions and also SSI, then I can also do that, but it will look a lot different, because it requires servers to interact as peers. They don't get to unilaterally declare what the state of the interaction is. Anybody can call them whenever they want; they don't necessarily start the workflow. And they must be able to call outward, too; an API that only documents behavior of half of the parties in the interaction isn't enough.

> Are you asserting that HTTP is incompatible with SSI.

Heavens no. But I am asserting that the narrow use of HTTP imagined in this API so far (where HTTP creates web service endpoints owned by institutional servers who play by different rules from their clients, and individuals have no way to play the issuer role without running web servers of their own) is incompatible with SSI as a cross-sovereign-domain interaction mechanism.

Reply:
> If we want an HTTP API that supports external interactions and also SSI, then I can also do that, but it will look a lot different, because it requires servers to interact as peers.

Ooooh, I think I get your objection now. This is *definitely not* defining an API designed for "external interactions". I agree that such an API would look very different from what we're talking about here.

In general, the API we're talking about here is meant to be "called internally" by some organization that wants to issue credentials from the organization. When I say "public API", I mean that the API may live on a different domain or subdomain from the institutions domain that is reachable via the public Internet (software as a service, but you still need to authz as the organization to issue credentials -- that is, it's still an "internal process" to the institution, but you're going across the public Internet to authz and hit the API). It's still only meant to be used by the institution as a part of it's credential issuing infrastructure.

To be clear, that's only what I think... it's starting to seem like others on this thread think it may be something else. So, we should confirm if everyone thinks what I said above.

Reply:
@David-Chadwick wrote:
> To my mind the first interface that should be defined is between the application services and the subject database. The rationale for this is that all issuing organisations (IdPs) already have subject databases, but dont have any VC issuing systems. So the first interface that needs to be built is between these two.

It took me a bit to parse this, but I think that's exactly what this API is attempting to do. See the comment above https://github.com/w3c-ccg/vc-issuer-http-api/pull/18#discussion_r387915866.

Reply:
@dhh1128 wrote:
> individuals have no way to play the issuer role without running web servers of their own

Yes, Digital Bazaar is 100% opposed to any ecosystem that *prevents* individuals from being able to play the role of issuer where they control the infrastructure... ideally, that infrastructure is under their control (code, software, hardware) and as inexpensive as possible (nothing is free, especially computing devices).

Reply:
@msporny. The same here. Verifiable Credentials Ltd is also 100% opposed to such an ecosystem.

Reply:
@msporny Referring to the architecture diagram, the first API that should be defined in my opinion is the arrow between (5) and (6). The rationale is that (6) is the only component that already exists in today's IdPs.

Reply:
> Ooooh, I think I get your objection now. This is definitely not defining an API designed for "external interactions" based on your definition of that above.

@msporny: Thanks for hanging with my flawed attempts until we discovered a place where our use of terms wasn't aligned. I think you may have teased out something important. Regarding the word "external" -- I'm not interested in whether a call originates inside or outside a firewall on the other side of the world. Like you, I agree that APIs can be locked down and protected according to circumstance, without changing the fundamental design. The external/internal distinction I was focused on is one about identity, not network, boundaries. If some entity inside Acme-the-issuer calls an Acme-the-issuer API from anywhere in the world, using whatever network connectivity it likes, it is always "internal" in the identity sense; the call is from an issuer's client to an issuer's service. In contrast, if an entity that is not "inside" Acme's identity calls the API (whether they do it while they're a guest on Acme's wifi, or from somewhere outside, then I was using the term "external" for that; it's "external" in the identity sense.

I'm happy to standardize an API for identity-internal clients to call. I just don't want it to be the same API that identity-external peers (not clients) call.

It sounds like there is less space between us than I thought -- and that David is also aligned with that. This makes me feel cheerful that we're getting somewhere. But I agree with your concern that others on this thread may see things differently. We should try to clarify further.

Reply:
I was also under the impression that "public"/"external" meant clients from different sovereign domains as the issuer's